crypto/src/hash/rescue/rp64_256_jive/mod.rs

```diff
@@ -0,0 +1,749 @@
+        // absorb the element into the rate portion of the state. we use 7-byte chunks because
+        // every 7-byte chunk is guaranteed to map to some field element.
+        let mut i = 0;
+        let mut buf = [0_u8; 8];
+        for (index, chunk) in bytes.chunks(7).enumerate() {
+            if index < num_elements - 1 {
+                buf[..7].copy_from_slice(chunk);
+            } else {
+                // if we are dealing with the last chunk, it may be smaller than 7 bytes long, so
+                // we need to handle it slightly differently. we also append a byte with value 1
+                // to the end of the string; this pads the string in such a way that adding
+                // trailing zeros results in different hash
+                let chunk_len = chunk.len();
+                buf = [0_u8; 8];
+                buf[..chunk_len].copy_from_slice(chunk);
+                buf[chunk_len] = 1;
+            }
+
+            // convert the bytes into a field element and absorb it into the rate portion of the
+            // state; if the rate is filled up, apply the Rescue permutation and start absorbing
+            // again from zero index.
+            state[RATE_RANGE.start + i] += BaseElement::new(u64::from_le_bytes(buf));
+            i += 1;
+            if i % RATE_WIDTH == 0 {
+                Self::apply_permutation(&mut state);
+                i = 0;
+            }
+        }
+
+        // if we absorbed some elements but didn't apply a permutation to them (would happen when
+        // the number of elements is not a multiple of RATE_WIDTH), apply the Rescue permutation.
+        // we don't need to apply any extra padding because we injected total number of elements
+        // in the input list into the capacity portion of the state during initialization.
+        if i > 0 {
+            Self::apply_permutation(&mut state);
+        }
+
+        // return the first 4 elements of the state as hash result
+        ElementDigest::new(state[DIGEST_RANGE].try_into().unwrap())
+    }
+
+    // We do not rely on the sponge construction to build our compression function. Instead, we use
+    // the Jive compression mode designed in https://eprint.iacr.org/2022/840.pdf.
+    fn merge(values: &[Self::Digest; 2]) -> Self::Digest {
+        // initialize the state by copying the digest elements into the state
+        let initial_state: [BaseElement; STATE_WIDTH] = Self::Digest::digests_as_elements(values)
+            .try_into()
+            .unwrap();
+        let mut state = initial_state;
+
+        // apply the Rescue permutation and apply the final Jive summation
+        Self::apply_permutation(&mut state);
+
+        let mut result = [BaseElement::ZERO; DIGEST_SIZE];
+        for (i, r) in result.iter_mut().enumerate() {
+            *r = initial_state[i]
+                + initial_state[DIGEST_SIZE + i]
+                + state[i]
+                + state[i + DIGEST_SIZE];
+        }
```

Comment:
Since the same code is also used on lines 228 - 234, we should probably put this into a stand-alone function. Maybe something like `apply_jive_summation()`?

Also minor nit: on line 197 we use `DIGEST_SIZE + i` but on line 199 we use `i + DIGEST_SIZE`. I'd probably make them consistent.

Reply:
`Since the same code is also used on lines 228 - 234, we should probably put this into a stand-alone function. Maybe something like apply_jive_summation()?`

Definitely. I had the same remark for MDS related functions (the fft2 / fft4 operations).